File: rust/nfts/src/states.rs
use chaindexing::augmenting_std::serde::{Deserialize, Serialize};
use chaindexing::states::{ContractState, StateMigrations};

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(crate = "chaindexing::augmenting_std::serde")]
pub struct Nft {
    pub token_id: u32,
    pub owner_address: String,
}

impl ContractState for Nft {
    fn table_name() -> &'static str {
        "nfts"
    }
}

pub struct NftMigrations;

impl StateMigrations for NftMigrations {
    fn migrations(&self) -> &'static [&'static str] {
        &["CREATE TABLE IF NOT EXISTS nfts (
                token_id INTEGER NOT NULL,
                owner_address TEXT NOT NULL
            )"]
    }
}

File: typescript/nfts/src/states.ts
import { BaseContractState, StateMigrations } from '../../../../chaindexing-ts/chaindexing-core/src';

export class Nft extends BaseContractState {
  constructor(
    public tokenId: number,
    public ownerAddress: string
  ) {
    super();
  }

  tableName(): string {
    return 'nfts';
  }
}

export class NftMigrations implements StateMigrations {
  migrations(): string[] {
    return [
      `CREATE TABLE IF NOT EXISTS nfts (
        token_id INTEGER NOT NULL,
        owner_address TEXT NOT NULL,
        created_at TIMESTAMP DEFAULT NOW(),
        updated_at TIMESTAMP DEFAULT NOW()
      )`,
      `CREATE UNIQUE INDEX IF NOT EXISTS nfts_token_id_index ON nfts(token_id)`,
    ];
  }
} 